### PR TITLE
Fix decorators skipping NCCL tests

### PR DESCRIPTION
**Pull request by Flamefire**
Avoid failures caused by tests exiting via sys.exit instead of `unittest.skip`

In particular it will not try to start the test (causing forks into subprocess) just to stop them (killing the subprocess) which is done in the test setup

Using `unittest.skip` decorators avoids the starting of the test in the first place.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

### CODE REVIEW

**Review by eqy — APPROVED** [review]
Does the same issue apply to other decorators that use the same `sys.exit` pattern e.g., `skip_if_no_gpu`?

**Comment by Flamefire** [comment]
> Does the same issue apply to other decorators that use the same `sys.exit` pattern e.g., `skip_if_no_gpu`?

Possibly yes, I'd strongly suggest to avoid this pattern. E.g. `exit_if_lt_x_gpu` could be replaced by `skipUnless`

`with_comms` in test_c10d_logger looks strange as it checks `BACKEND` which seems to be a constant.

`exit_if_lt_x_gpu` & `with_comms` in test_functional_api seems to duplication (both check GPUs against world size)

However many of the other usages of `sys.exit(TEST_SKIPS...` involve the "world size", e.g. usages of `exit_if_lt_x_gpu` which I think is not possible to do in a decorator as either `self` is not available until the method is called or `$WORLD_SIZE` might not be set until the test is started spawing processes.

I did what I could now to reduce the issues and do some cleanup.

File: torch/testing/_internal/common_distributed.py [hunk 1]
```diff
@@ -102,21 +102,22 @@ class DistTestCases:
...
     test is run, each rank has its own GPU via ``torch.cuda.device(rank)``."""
 
     @wraps(func)
     def wrapper(*args, **kwargs):
-        if not torch.cuda.is_available():
-            sys.exit(TEST_SKIPS["no_cuda"].exit_code)
-        world_size = int(os.environ["WORLD_SIZE"])
-        if torch.cuda.device_count() < world_size:
-            sys.exit(TEST_SKIPS[f"multi-gpu-{world_size}"].exit_code)
+        exit_if_lt_x_gpu(int(os.environ["WORLD_SIZE"]))
```
Comment (wconstab) [inline torch/testing/_internal/common_distributed.py:116]: can you comment on the state of this code? is this still 'ok' to use exit or is this just an area that needs further work and not handled in this PR?

File: test/distributed/test_c10d_logger.py [hunk 2]
```diff
@@ -39,8 +39,8 @@ def with_comms(func=None):
 
     @wraps(func)
     def wrapper(self, *args, **kwargs):
-        if BACKEND == dist.Backend.NCCL and torch.cuda.device_count() < self.world_size:
-            sys.exit(TEST_SKIPS[f"multi-gpu-{self.world_size}"].exit_code)
+        if BACKEND == dist.Backend.NCCL:
+            exit_if_lt_x_gpu(self.world_size)
```
Comment (wconstab) [inline test/distributed/test_c10d_logger.py:43]: similar comment from below- unclear if using exit here is supposed to be OK or if this is future cleanup work too

File: torch/testing/_internal/common_distributed.py (hunk 1 above)
Comment (Flamefire) [reply torch/testing/_internal/common_distributed.py:116]: I tried to explain this in https://github.com/pytorch/pytorch/pull/122397#issuecomment-2014941519

In short: It needs further work and might not even be possible.

Question here is when exactly `$WORLD_SIZE` is set. I.e. is this available when the decorator is executed or only when the function is executed, i.e. after the test setup (likely involving forking). In the latter case this cannot be helped.

Similar below: Dependency on `self.world_size` and we don't have `self` available in the decorator. And especially as this property can be set/changed by subclasses I figure it is very hard to fix this.

However what could be improved is the duplication:
- `with_comms` is defined at least twice and very similarly
- Unclear differences between usage of `self.world_size` and `$WORLD_SIZE`, maybe use either one?
- After the previous point maybe a (new) separate decorator `require_gpus_foreach_rank` could be used to move this out of `with_comms`
- In all those cases we can skip early (`unittest.skip`) if there are no GPUs available at all. Might already workaround a for common case (CPU only machine)

**Comment by Flamefire** [comment]
@eqy @wconstab can this be merged please?

**Comment by eqy** [comment]
No objects from me but I assume it needs a stamp with someone with merge rights

**Review by wconstab — APPROVED** [review]
lgtm.

could you also comment on the future work that you would propose?

I think these changes look correct, but I noticed that only a small set of tests (4) used the new 'skip' functionality

**Comment by Flamefire** [comment]
> could you also comment on the future work that you would propose?

I'd propose to double-check the uses of the `exit_if_*` markers/functions whether the condition can be determined before the test starts. As mentioned above this is hard if the condition is `self.worldsize` which is determined by env variables or only set after the test was initialized. Similar other cases exist (e.g. `os.environ["WORLD_SIZE"]` which IIRC is set when the program is forked for starting the test)

> I think these changes look correct, but I noticed that only a small set of tests (4) used the new 'skip' functionality

That only looks like this, see the first commit which has [the most important changes](https://github.com/pytorch/pytorch/pull/122397/files#diff-922fddd5e2265ab3b10b0f0741cb937b14696c52266a0ca9279102657991441bL175-R160)

I.e. instead of starting the function and conditionally exit the process when not enough GPUs are available skip the test without starting it

**Comment by Flamefire** [comment]
@wconstab The failure seems unrelated to these changes and only happened in the merge attempt but not before. Might be a regression? `main` looks pretty red right now

**Comment by Flamefire** [comment]
@eqy @wconstab can this be reopened and merged please?

**Comment by wconstab** [comment]
Yea, I'm not sure what happened. Let me also pull in @fegin and @mori360  who are looking into some test changes.

**Pull request closed without merging**